### PR TITLE
Implemented storage encryption feature support

### PR DESCRIPTION
**Pull request by kasabe28**
# Proposed Changes

- Add new optional field encryption with single attribute secretRef in VolumeSpec
- Add validations to check secretRef attribute is immutable and secretRef.name is not empty.
- Modify volume-test.go to add encryption field testcases
- Modify api reference docs using gen-crd-api-reference-docs

### CODE REVIEW

File: onmetal-apiserver/internal/apis/storage/validation/volume.go [hunk 1]
```diff
@@ -118,6 +125,9 @@ func validateVolumeSpecUpdate(newSpec, oldSpec *storage.VolumeSpec, fldPath *fie
 
 	allErrs = append(allErrs, onmetalapivalidation.ValidateImmutableField(newSpec.VolumeClassRef, oldSpec.VolumeClassRef, fldPath.Child("volumeClassRef"))...)
 	allErrs = append(allErrs, onmetalapivalidation.ValidateSetOnceField(newSpec.VolumePoolRef, oldSpec.VolumePoolRef, fldPath.Child("volumePoolRef"))...)
+	if newSpec.Encryption != nil {
+		allErrs = append(allErrs, onmetalapivalidation.ValidateImmutableField(newSpec.Encryption.SecretRef, oldSpec.Encryption.SecretRef, fldPath.Child("encryption").Child("secretRef"))...)
+	}
```
Comment (adracus) [inline onmetal-apiserver/internal/apis/storage/validation/volume.go:130]: The entire `encryption` field should be immutable - it should not be possible to add encryption to a volume after it has been created.

**Review by adracus — CHANGES_REQUESTED** [review]

File: api/storage/v1alpha1/volume_types.go [hunk 2]
```diff
@@ -52,6 +59,9 @@ type VolumeSpec struct {
 	// Tolerations define tolerations the Volume has. Only any VolumePool whose taints
 	// covered by Tolerations will be considered to host the Volume.
 	Tolerations []commonv1alpha1.Toleration `json:"tolerations,omitempty"`
+	// Encryption is an optional field to specify encryption key secret reference to encrypt a Volume.
+	// This secret is created by user with encryptionKey as Key and base64 encoded 256-bit encryption key as Value.
```
[the remark this answers is not in the document]
Comment (kasabe28) [reply api/storage/v1alpha1/volume_types.go]: Agree. Done.

File: api/storage/v1alpha1/volume_types.go [hunk 3]
```diff
@@ -29,6 +30,12 @@ var VolumeGK = schema.GroupKind{
 	Kind:  "Volume",
 }
 
+// VolumeEncryption represents information to encrypt a volume.
+type VolumeEncryption struct {
+	// SecretRef references the Secret containing the encryption key to encrypt a Volume.
+	SecretRef *corev1.LocalObjectReference `json:"secretRef,omitempty"`
```
[the remark this answers is not in the document]
Comment (kasabe28) [reply api/storage/v1alpha1/volume_types.go]: Agree. Done

File: onmetal-apiserver/internal/apis/storage/validation/volume.go (hunk 1 above)
Comment (kasabe28) [reply onmetal-apiserver/internal/apis/storage/validation/volume.go]: Agree. Done

File: onmetal-apiserver/internal/apis/storage/validation/volume_test.go [hunk 4]
```diff
@@ -181,5 +191,20 @@ var _ = Describe("Volume", func() {
 			},
 			Not(ContainElement(ImmutableField("spec.volumePoolRef"))),
 		),
+		Entry("immutable encryption",
```
Comment (adracus) [inline onmetal-apiserver/internal/apis/storage/validation/volume_test.go:194]: There should be a test for changing the entire `encryption` field.

**Review by adracus — CHANGES_REQUESTED** [review]

File: api/storage/v1alpha1/volume_types.go [hunk 5]
```diff
@@ -52,6 +59,9 @@ type VolumeSpec struct {
 	// Tolerations define tolerations the Volume has. Only any VolumePool whose taints
 	// covered by Tolerations will be considered to host the Volume.
 	Tolerations []commonv1alpha1.Toleration `json:"tolerations,omitempty"`
+	// Encryption is an optional field.
+	// Encryption makes a Volume encrypted by the supplied attributes.
```
[the remark this answers is not in the document]
Comment (kasabe28) [reply api/storage/v1alpha1/volume_types.go]: Done

File: onmetal-apiserver/internal/apis/storage/validation/volume.go [hunk 6]
```diff
@@ -118,6 +125,9 @@ func validateVolumeSpecUpdate(newSpec, oldSpec *storage.VolumeSpec, fldPath *fie
 
 	allErrs = append(allErrs, onmetalapivalidation.ValidateImmutableField(newSpec.VolumeClassRef, oldSpec.VolumeClassRef, fldPath.Child("volumeClassRef"))...)
 	allErrs = append(allErrs, onmetalapivalidation.ValidateSetOnceField(newSpec.VolumePoolRef, oldSpec.VolumePoolRef, fldPath.Child("volumePoolRef"))...)
+	if newSpec.Encryption != nil {
+		allErrs = append(allErrs, onmetalapivalidation.ValidateImmutableField(newSpec.Encryption, oldSpec.Encryption, fldPath.Child("encryption"))...)
+	}
```
[the remark this answers is not in the document]
Comment (kasabe28) [reply onmetal-apiserver/internal/apis/storage/validation/volume.go]: Removed

File: onmetal-apiserver/internal/apis/storage/validation/volume_test.go (hunk 4 above)
Comment (kasabe28) [reply onmetal-apiserver/internal/apis/storage/validation/volume_test.go]: Added testcases to Add/Remove entire `encryption` field

File: onmetal-apiserver/internal/apis/storage/volume_types.go [hunk 7]
```diff
@@ -29,6 +30,12 @@ var VolumeGK = schema.GroupKind{
 	Kind:  "Volume",
 }
 
+// VolumeEncryption represents information to encrypt a volume.
+type VolumeEncryption struct {
+	// SecretRef references the Secret containing the encryption key to encrypt a Volume.
```
[the remark this answers is not in the document]
Comment (kasabe28) [reply onmetal-apiserver/internal/apis/storage/volume_types.go:35]: Done

File: docs/api-reference/compute.md [hunk 8]
```diff
@@ -113,6 +113,20 @@ If empty, a scheduler will figure out an appropriate pool to run the machine in.
...
 <td>
+<code>power</code><br/>
+<em>
+<a href="#compute.api.onmetal.de/v1alpha1.Power">
+Power
+</a>
+</em>
+</td>
+<td>
+<p>Power ist the desired machine power state.
```
Comment (ushabelgur) [inline docs/api-reference/compute.md:124]: Typo

File: onmetal-apiserver/internal/apis/storage/validation/volume.go [hunk 9]
```diff
@@ -100,6 +101,12 @@ func validateVolumeSpec(spec *storage.VolumeSpec, fldPath *field.Path) field.Err
 		}
 	}
 
+	if spec.Encryption != nil {
+		for _, msg := range apivalidation.NameIsDNSLabel(spec.Encryption.SecretRef.Name, false) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("encryption").Child("secretRef").Child("name"), spec.Encryption.SecretRef.Name, msg))
+		}
+	}
```
Comment (ushabelgur) [inline onmetal-apiserver/internal/apis/storage/validation/volume.go:108]: should we add null validation check for secreteRef.Data encryptionKey field?

File: docs/api-reference/compute.md (hunk 8 above)
Comment (kasabe28) [reply docs/api-reference/compute.md]: Done

File: onmetal-apiserver/internal/apis/storage/validation/volume.go (hunk 9 above)
Comment (kasabe28) [reply onmetal-apiserver/internal/apis/storage/validation/volume.go:108]: We cant add it here as secret is not available.

**Review by adracus — APPROVED** [review]

**Pull request merged**